File: Example/Realm/SyncKitRealmExample/QSSharedCompanyTableViewController.m
//
//  QSSharedCompanyTableViewController.m
//  SyncKitRealm
//
//  Created by Manuel Entrena on 04/05/2017.
//  Copyright © 2017 Colourbox. All rights reserved.
//

#import "QSSharedCompanyTableViewController.h"
#import "QSEmployeeTableViewController.h"

#import <Realm/Realm.h>

#import "QSCompanyTableViewCell.h"
#import "QSCompany.h"
@import SyncKit;

@interface QSSharedCompanyTableViewController () <UICloudSharingControllerDelegate, QSMultiRealmResultsControllerDelegate>

@property (nonatomic, strong) RLMNotificationToken *notificationToken;

@property (nonatomic, weak) IBOutlet UIButton *syncButton;
@property (nonatomic, weak) IBOutlet UIActivityIndicatorView *indicatorView;

@property (nonatomic, strong) QSMultiRealmResultsController *resultsController;

@property (nonatomic, strong) QSCompany *sharingCompany;

@end

@implementation QSSharedCompanyTableViewController

- (void)awakeFromNib
{
    [super awakeFromNib];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    [self setupResultsController];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    self.clearsSelectionOnViewWillAppear = self.splitViewController.isCollapsed;
    
    [self.tableView reloadData];
}

- (void)setupResultsController
{
    if (!_resultsController) {
        _resultsController = [self.synchronizer multiRealmResultsControllerWithClass:[QSCompany class] predicate:nil];
        _resultsController.delegate = self;
    }
}

- (void)multiRealmResultsControllerDidChangeRealms:(QSMultiRealmResultsController *)controller
{
    [self.tableView reloadData];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)didTapSynchronize:(id)sender
{
    [self synchronizeWithCompletion:nil];
}

#pragma mark - Segues

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"showEmployees"]) {
        QSCompany *company = (QSCompany *)sender;
        QSEmployeeTableViewController *controller = (QSEmployeeTableViewController *)segue.destinationViewController;
        controller.realm = company.realm;
        controller.company = company;
        CKShare *share = [self.synchronizer shareFor:company];
        controller.canWrite = share.currentUserParticipant.permission == CKShareParticipantPermissionReadWrite;;
        controller.navigationItem.leftBarButtonItem = self.splitViewController.displayModeButtonItem;
        controller.navigationItem.leftItemsSupplementBackButton = YES;
    }
}


#pragma mark - Table View

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return self.resultsController.results.count;
}


- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    RLMResults *sectionResults = self.resultsController.results[section];
    return sectionResults.count;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    QSCompanyTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell" forIndexPath:indexPath];
    
    
    QSCompany *company = [self companyAtIndexPath:indexPath];
    
    cell.nameLabel.text = company.name;
    
    CKShare *share = [self.synchronizer shareFor:company];
    if (share) {
        [cell.sharingButton setTitle:@"Sharing" forState:UIControlStateNormal];
    } else {
        [cell.sharingButton setTitle:@"Share" forState:UIControlStateNormal];
    }
    
    __weak QSSharedCompanyTableViewController *weakSelf = self;
    
    cell.shareButtonAction = ^{
        [weakSelf shareCompany:company];
    };
    
    return cell;
}

- (QSCompany *)companyAtIndexPath:(NSIndexPath *)indexPath
{
    RLMResults *sectionResults = self.resultsController.results[indexPath.section];
    QSCompany *company = sectionResults[indexPath.row];
    return company;
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the specified item to be editable.
    return YES;
}


- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        QSCompany *company = [self companyAtIndexPath:indexPath];
        [company.realm transactionWithBlock:^{
            [company.realm deleteObject:company];
        }];
    }
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    QSCompany *company = [self companyAtIndexPath:indexPath];
    [self performSegueWithIdentifier:@"showEmployees" sender:company];
    
    [self.tableView deselectRowAtIndexPath:indexPath animated:YES];
}

#pragma mark - Sync

- (void)synchronizeWithCompletion:(void(^)(NSError *error))completion
{
    [self showLoading:YES];
    __weak QSSharedCompanyTableViewController *weakSelf = self;
    [self.synchronizer synchronizeWithCompletion:^(NSError *error) {
        
        [weakSelf showLoading:NO];
        
        if (error) {
            UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Error" message:[NSString stringWithFormat:@"Error: %@", error] preferredStyle:UIAlertControllerStyleAlert];
            
            [alertController addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil]];
            
            [weakSelf presentViewController:alertController animated:YES completion:nil];
        }
        
        [self.tableView reloadData];
        
        if (completion) {
            completion(error);
        }
    }];
}

#pragma mark - Sharing

- (void)shareCompany:(QSCompany *)company
{
    self.sharingCompany = company;
    
    __weak QSSharedCompanyTableViewController *weakSelf = self;
    
    [self synchronizeWithCompletion:^(NSError *error) {
        
        if (!error) {
            UICloudSharingController *sharingController;
            CKShare *share = [weakSelf.synchronizer shareFor:company];
            CKContainer *container = [CKContainer containerWithIdentifier:weakSelf.synchronizer.containerIdentifier];
            if (share) {
                sharingController = [[UICloudSharingController alloc] initWithShare:share container:container];
            } else {
                sharingController = [[UICloudSharingController alloc] initWithPreparationHandler:^(UICloudSharingController * _Nonnull controller, void (^ _Nonnull preparationCompletionHandler)(CKShare * _Nullable, CKContainer * _Nullable, NSError * _Nullable)) {
                    
                    [self.synchronizer shareWithObject:company publicPermission:CKShareParticipantPermissionReadOnly participants:@[] completion:^(CKShare *share, NSError *error) {
                        
                        share[CKShareTitleKey] = company.name;
                        preparationCompletionHandler(share, container, error);
                    }];
                }];
            }
            
            sharingController.availablePermissions = UICloudSharingPermissionAllowPublic | UICloudSharingPermissionAllowReadOnly | UICloudSharingPermissionAllowReadWrite;
            sharingController.delegate = self;
            [weakSelf presentViewController:sharingController animated:YES completion:nil];
        }
    }];
}

#pragma mark - UICloudSharingControllerDelegate

- (NSString *)itemTitleForCloudSharingController:(UICloudSharingController *)csc
{
    return self.sharingCompany.name;
}

- (void)cloudSharingController:(UICloudSharingController *)csc failedToSaveShareWithError:(NSError *)error
{
    NSLog(@"Error saving: %@", error);
}

- (void)cloudSharingControllerDidSaveShare:(UICloudSharingController *)csc
{
    [self.synchronizer saveShare:csc.share for:self.sharingCompany];
    [self.synchronizer synchronizeWithCompletion:nil];
}

- (void)cloudSharingControllerDidStopSharing:(UICloudSharingController *)csc
{
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [self synchronizeWithCompletion:nil];
    });
}

#pragma mark - Loading

- (void)showLoading:(BOOL)loading
{
    if (loading) {
        self.syncButton.hidden = YES;
        [self.indicatorView startAnimating];
    } else {
        self.syncButton.hidden = NO;
        [self.indicatorView stopAnimating];
    }
}

@end

File: Example/RealmSwift/SyncKitRealmSwift/QSSharedCompanyTableViewController.swift
//
//  QSSharedCompanyTableViewController.swift
//  SyncKitRealmSwiftExample
//
//  Created by Manuel Entrena on 08/06/2018.
//  Copyright © 2018 Manuel Entrena. All rights reserved.
//

import UIKit
import SyncKit

class QSSharedCompanyTableViewController: UITableViewController, UICloudSharingControllerDelegate {
    
    var synchronizer: QSCloudKitSynchronizer!
    
    var sharingCompany: QSCompany?
    
    var resultsController: MultiRealmResultsController<QSCompany>!
    
    @IBOutlet weak var syncButton: UIButton?
    @IBOutlet weak var indicatorView: UIActivityIndicatorView?
    
    typealias RealmType = QSCompany
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupResultsController()
    }
    
    func setupResultsController() {
        
        resultsController = synchronizer.multiRealmResultsController()
        resultsController.didChangeRealms = { [weak self] controller in
            self?.tableView.reloadData()
        }
    }
    
    @IBAction func didTapSynchronize() {
        
        synchronize(completion: nil)
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        
        if segue.identifier == "showEmployees" {
            
            let company = sender as! QSCompany
            
            let controller = segue.destination as! QSEmployeeTableViewController
            controller.realm = company.realm
            controller.company = company
            let share = synchronizer.share(for: company)
            controller.canWrite = (share?.currentUserParticipant?.permission ?? .readOnly) == .readWrite;
            controller.navigationItem.leftBarButtonItem = splitViewController?.displayModeButtonItem
            controller.navigationItem.leftItemsSupplementBackButton = true
        }
    }
    
    // MARK: - TableView
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        return resultsController.results.count
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        
        let results = resultsController.results[section]
        return results.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let cell = tableView.dequeueReusableCell(withIdentifier: "cell") as! CompanyCell
        
        let company = getCompany(at: indexPath)
        
        cell.nameLabel?.text = company?.name
        
        if let _ = synchronizer.share(for: company!) {
            cell.sharingButton?.setTitle("Sharing", for: .normal)
        } else {
            cell.sharingButton?.setTitle("Share", for: .normal)
        }
        
        cell.shareButtonAction = { [weak self] in
            self?.shareCompany(company!)
        }
        
        return cell
    }
    
    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        
        return true
    }
    
    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
        
        if editingStyle == .delete {
            
            let company = getCompany(at: indexPath)
            try! company!.realm?.write {
                company!.realm!.delete(company!)
            }
        }
    }
    
    func getCompany(at indexPath: IndexPath) -> QSCompany! {
        
        let results = resultsController.results[indexPath.section]
        return results[indexPath.row]
    }

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        
        let company = getCompany(at: indexPath)
        performSegue(withIdentifier: "showEmployees", sender: company)
        
        tableView.deselectRow(at: indexPath, animated: true)
    }
    
    // MARK: - Sync
    
    func synchronize(completion: ((Error?)->())?) {
        
        showLoading(true)
        synchronizer.synchronize { [weak self] (error) in
            
            self?.showLoading(false)
            
            if let error = error {
                
                let alertController = UIAlertController(title: "Error", message: "Error: \(error.localizedDescription)", preferredStyle: .alert)
                alertController.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
                
                self?.present(alertController, animated: true, completion: nil)
                
            }
            
            self?.tableView.reloadData()
            
            completion?(error)
        }
    }
    
    // MARK: - Sharing
    
    func shareCompany(_ company: QSCompany) {
        
        sharingCompany = company
        
        synchronize { [weak self] (error) in
            if error == nil,
                let strongSelf = self {
                
                let sharingController: UICloudSharingController
                let share = strongSelf.synchronizer.share(for: company)
                let container = CKContainer(identifier: strongSelf.synchronizer.containerIdentifier)
                
                if let share = share {
                    sharingController = UICloudSharingController(share: share, container: container)
                } else {
                    sharingController = UICloudSharingController(preparationHandler: { (controller, handler) in
                        
                        strongSelf.synchronizer.share(object: company,
                                                      publicPermission: .readOnly,
                                                      participants: [],
                                                      completion: { (share, error) in
                                                        
                                                        share?[CKShareTitleKey] = company.name as NSString?
                                                        handler(share, container, error)
                        })
                    })
                }
                
                sharingController.availablePermissions = [.allowPublic, .allowReadOnly, .allowPrivate]
                sharingController.delegate = self
                
                strongSelf.present(sharingController, animated: true, completion: nil)
            }
        }
    }
    
    // MARK : - UICloudSharingControllerDelegate
    
    func itemTitle(for csc: UICloudSharingController) -> String? {
        
        return sharingCompany?.name
    }
    
    func cloudSharingControllerDidSaveShare(_ csc: UICloudSharingController) {
        
        guard let company = sharingCompany,
            let share = csc.share else {
                return
        }
        synchronizer.saveShare(share, for: company)
        synchronize(completion: nil)
    }
    
    func cloudSharingControllerDidStopSharing(_ csc: UICloudSharingController) {
        
        DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
            self.synchronize(completion: nil)
        }
    }
    
    func cloudSharingController(_ csc: UICloudSharingController, failedToSaveShareWithError error: Error) {
        
        print("Failed to save share: \(error.localizedDescription)")
    }
    
    // MARK: - Loading
    
    func showLoading(_ loading: Bool) {
        
        if loading {
            syncButton?.isHidden = true
            indicatorView?.startAnimating()
        } else {
            syncButton?.isHidden = false
            indicatorView?.stopAnimating()
        }
    }
}
